libeos-payg/util.c

```diff
@@ -0,0 +1,247 @@
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Alternatively, the contents of this file may be used under the terms of the
+ * GNU Lesser General Public License Version 2.1 or later (the "LGPL"), in
+ * which case the provisions of the LGPL are applicable instead of those above.
+ * If you wish to allow use of your version of this file only under the terms
+ * of the LGPL, and not to allow others to use your version of this file under
+ * the terms of the MPL, indicate your decision by deleting the provisions
+ * above and replace them with the notice and other provisions required by the
+ * LGPL. If you do not delete the provisions above, a recipient may use your
+ * version of this file under the terms of either the MPL or the LGPL.
+ */
+
+#include <libeos-payg/util.h>
+#include <efivar.h>
+
+#define EFI_GLOBAL_VARIABLE_GUID EFI_GUID(0x8be4df61, 0x93ca, 0x11d2, 0xaa0d, 0x00, 0xe0, 0x98, 0x03, 0x2b, 0x8c)
+
+/**
+ * EpgDebugFlags:
+ *
+ * Flags set by the EFI variable EOSPAYG_debug which affect the behavior of
+ * eos-paygd. For an explanation of each, see
+ * https://phabricator.endlessm.com/w/software/pay-as-you-go/
+ *
+ * If adding to this enum, update _read_eospayg_debug() and if necessary extend
+ * it to work with more than one byte.
+ */
+typedef enum {
+  EPG_DEBUG_SET_G_MESSAGES_DEBUG_ALL = 1 << 0,
+  EPG_DEBUG_SECURE_BOOT_OFF = 1 << 1,
+  EPG_DEBUG_SECURE_BOOT_ON = 1 << 2,
+  EPG_DEBUG_EOSPAYG_ACTIVE_OFF = 1 << 3,
+  EPG_DEBUG_EOSPAYG_ACTIVE_ON = 1 << 4,
+  EPG_DEBUG_DONT_USE_WATCHDOG = 1 << 5,
+  EPG_DEBUG_DONT_USE_LSM = 1 << 6,
+  EPG_DEBUG_SKIP_SECURITYLEVEL_CHECK = 1 << 7,
+} EpgDebugFlags;
+
+static EpgDebugFlags
+_read_eospayg_debug (void)
+{
+  uint8_t *debug_efivar;
+  size_t data_size;
+  uint32_t attributes;
+  int ret;
+  EpgDebugFlags debug_flags = 0;
+
+  if (efi_get_variable_exists (EOSPAYG_GUID, "EOSPAYG_debug") < 0)
+    return debug_flags;
+
+  /* For now we only look at the first byte but let's allow it to be bigger so
+   * it can be extended in the future.
+   */
+  ret = efi_get_variable (EOSPAYG_GUID, "EOSPAYG_debug", &debug_efivar, &data_size, &attributes);
+  if (ret < 0 || !debug_efivar || data_size < 1)
+    {
+      g_printerr ("%s: Failed to read EOSPAYG_debug\n", G_STRFUNC);
+      return debug_flags;
+    }
+
+  g_debug ("%s: EOSPAYG_debug is set to %u", G_STRFUNC, debug_efivar[0]);
+
+  /* The C standard doesn't guarantee the underlying type of an enum, so let's
+   * convert manually to be safe.
+   */
+  if (debug_efivar[0] & EPG_DEBUG_SET_G_MESSAGES_DEBUG_ALL)
+    debug_flags |= EPG_DEBUG_SET_G_MESSAGES_DEBUG_ALL;
+  if (debug_efivar[0] & EPG_DEBUG_SECURE_BOOT_OFF)
+    debug_flags |= EPG_DEBUG_SECURE_BOOT_OFF;
+  if (debug_efivar[0] & EPG_DEBUG_SECURE_BOOT_ON)
+    debug_flags |= EPG_DEBUG_SECURE_BOOT_ON;
```

Comment:
This is exactly what I wanted to see, where secure boot can be forced on, or force off, or just left alone.